### ISSUE TITLE
Feature request fail2ban

### ISSUE THREAD

**Issue by km4efp** [gogits/gogs#2384]
Enable user auth fail log for use with fail2ban in preventing brute force attacks

**Comment by Unknwon** [comment]
Thanks your feedback!

> Enable user auth fail log for use with fail2ban

How does this work?

**Comment by bkcsoft** [comment]
@Unknwon Basically the authentication route has to log failed login-attempts to file with the offending IP
http://www.fail2ban.org/wiki/index.php/Main_Page

actually it would be simplest if Macaron just printed the ip-address to the log as-well :smile:

**Comment by Unknwon** [comment]
Thanks @bkcsoft !

Sounds like a simple integration (log-level) to fail2ban?

**Comment by bkcsoft** [comment]
@Unknwon well, it's a simple integration, _if_ the logs from gogs printed out ip-addresses ;)

**Comment by Unknwon** [comment]
Only IP addresses? To any file?

**Comment by bkcsoft** [comment]
well, fail2ban needs a status-code and an IP... preferably turn this:
```
gogs | [Macaron] Completed /img/favicon.png 304 Not Modified in 91.047µs
```
into:
```
gogs | [Macaron] Completed /img/favicon.png 304 Not Modified in 91.047µs from 172.16.0.14
```

**Comment by Unknwon** [comment]
Thanks. Does it have any so-called standard form of log?

**Comment by wqweto** [comment]
@Unknwon `fail2ban` uses flexible set of filtering rules based on filter/ignore regexes e.g. something like this
```
failregex = (?i)[Macaron] Failed /login .* from <HOST>
```
. . . can be used to filter failed logon attempts *and* gather IP through \<HOST\> capture.

**Comment by Unknwon** [comment]
Thanks @wqweto !

**Comment by Unknwon** [comment]
I think it is now as easy as adding a [NewLogger](https://github.com/gogits/gogs/blob/master/modules%2Flog%2Flog.go#L22) with `file` type and filename `fail2ban.log`, and record failures in https://github.com/gogits/gogs/blob/master/routers%2Fuser%2Fauth.go#L63

**Comment by AndreaGiardini** [comment]
I would love to see this in Gogs! I subscribe :+1:

**Comment by AndreaGiardini** [comment]
I think #2291 is a duplicate

**Comment by Unknwon** [comment]
@AndreaGiardini I think this one is the subset of #2291...

**Comment by masupilamie** [comment]
I would also like to be able to use fail2ban with gogs

**Comment by AndreaGiardini** [comment]
@Unknwon Why not logging directly the failed attempts in gogs.log?

**Comment by masupilamie** [comment]
I think #2291 would be a very good addition to gogs. But logging a failed attempt with ip has its perks also: fail2ban has the ability to block (or redirect) ports so the "attacker" ip can be completely cut off, plus e-mail notifications out of the box.

**Comment by nobecutan** [comment]
+1

**Comment by zerowaitstate** [comment]
The inability to know whether bots are failing login attempts is a deal-breaker for anyone using this for private repos.  I set up a gogs site and within a week it was being crawled by shodan.io.  It's frustrating having to set up a WAF simply because gogs doesn't record login attempts.

**Comment by masupilamie** [comment]
Tested yweber's commit and it seems to work, would be nice to see this getting merged so it can make its way into the binary releases.

Fail2ban jail configuration used (added to /etc/fail2ban/jail.conf):
```[gogs]
enabled  = true
logpath = /home/git/go/src/github.com/gogits/gogs/log/gogs.log
port = 3000
filter  = gogs
```

Fail2ban filter used (created at /etc/fail2ban/filter.d/gogs.conf):
```
[Definition]

failregex = <HOST> Authentication failure for user
            <HOST> Authentication failure using wrong token

ignoreregex =
```

**Comment by Kambaa** [comment]
@masupilamie i would like to implement this too, does the configuation of the gogs setup is the same as the installation placed on the gogs website? thank you in advance...

**Comment by masupilamie** [comment]
@Kambaa Because the commits made by yweber (and me) are not (yet) in the official gogs git repository, you'll have to manually compile gogs after adding (cherry-pick) those commits to your local git repository (something like is explained here: http://voidcanvas.com/cherry-pick-format-patch/ ). Gogs itself does not need any extra configuration after this, it will log failed attempts by default. Of course you'll need to configure fail2ban when the loggin works (see my example above).

**Comment by quux** [comment]
Hoping those commits make it in an official release soon.

Meanwhile, if this could help somebody, I have a working solution using Nginx as a reverse-proxy to Gogs. 

I have a fail2ban filter watching `/var/log/nginx/access.log` for POSTs to `/user/login` with a 200 HTTP status code. (Successful authentication returns a 302...)

`failregex = ^<HOST> -.*"POST /user/login[^"]+" 200.*$`

**Comment by webracer999** [comment]
+1 for fail2ban support

**Comment by machsix** [comment]
Is there any progress with this issue? Thank you!

**Comment by CWempe** [comment]
I am also interested.
And I am wondering if this discussion is only about HTTP or SSH, too.

**Comment by tschallacka** [comment]
Is this already implemented or is it still not here?
+1

**Comment by angelmartinezn** [comment]
Hi, I did configure fail2ban to work with gogs (over an apache proxy). You'll find the configuration files for fail2ban [here](https://github.com/angelmartinezn/fail2ban-apache-gogs-proxy)

**Comment by webracer999** [comment]
Hello @angelmartinezn,

thank you for your fail2ban configuration, but I think there's an uncovered scenario:
You can also access data (and login) via git command. 

A "git pull" produces on my gogs installation two http requests on
GET (...)/info/refs?service=git-upload-pack

The first request is responded with an 401 http code. After that, providing the password, it gets an 200.

So an attacker can still brute force on this URL. :-(

I don't know, if there are other scenarios ...
So the best thing would be, if fail2ban could log incorrect logins.

Regards, webracer999

**Comment by angelmartinezn** [comment]
> 
> 
> Hello @angelmartinezn,
> 
> thank you for your fail2ban configuration, but I think there's an uncovered scenario:
> You can also access data (and login) via git command.
> 
> A "git pull" produces on my gogs installation two http requests on
> GET (...)/info/refs?service=git-upload-pack
> 
> The first request is responded with an 401 http code. After that, providing the password, it gets an 200.
> 
> So an attacker can still brute force on this URL. :-(
> 
> I don't know, if there are other scenarios ...
> So the best thing would be, if fail2ban could log incorrect logins.
> 
> Regards, webracer999

You're absolutely right, @webracer999. I agree with you in your conclusion of the best way to get this is to have full support from gogs.

Anyway, I updated the fail2ban filter, feel free to try it (I've tested it with git pull authentication method and it works well in my setup).

Regards,

**Comment by tschallacka** [comment]
Welcome to 2021 What is the status on this issue? Is it being worked on or resolved in another update and can this be closed?

**Comment by webracer999** [comment]
BTW I've enabled TFA in gogs. Today, I think it's the better solution. Fail2Ban fails, when an attacker uses multiple IP addresses ...

**Comment by Schlumpf9** [comment]
+1 for better logging

**Comment by YeFei572** [comment]
+1 for better logging

**Comment by YeFei572** [comment]
![image](https://user-images.githubusercontent.com/26360338/147755193-8d66e5d3-2f7b-4eb9-add0-03b4f1c5b3d5.png)
